adapters/src/transport/mod.rs

```diff
@@ -0,0 +1,110 @@
+use anyhow::{Error as AnyError, Result as AnyResult};
+use once_cell::sync::Lazy;
+use serde_yaml::Value as YamlValue;
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+
+mod file;
+
+pub use file::FileInputTransport;
+
+/// Static map of supported input transports.
+// TODO: support for registering new transports at runtime in order to allow
+// external crates to implement new transports.
+static INPUT_TRANSPORT: Lazy<BTreeMap<&'static str, Box<dyn InputTransport>>> = Lazy::new(|| {
+    BTreeMap::from([(
+        "file",
+        Box::new(FileInputTransport) as Box<dyn InputTransport>,
+    )])
+});
+
+/// Trait that represents a specific data transport.
+///
+/// This is a factory trait that creates transport endpoints for a specific
+/// data format.
+pub trait InputTransport: Send + Sync {
+    /// Unique name of the data transport.
+    fn name(&self) -> Cow<'static, str>;
+
+    /// Create a new transport endpoint.
+    ///
+    /// Create and initializes a transport endpoint.  The endpoint will push
+    /// received data to the provided input consumer.  The endpoint is created
+    /// in a paused state.
+    ///
+    /// # Arguments
+    ///
+    /// * `config` - Transport-specific configuration.
+    ///
+    /// * `consumer` - Input consumer that will receive data from the endpoint.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the specified configuration is invalid or the endpoint failed
+    /// to initialize (e.g., the endpoint was not able to establish a network
+    /// connection).
+    fn new_endpoint(
+        &self,
+        config: &YamlValue,
+        consumer: Box<dyn InputConsumer>,
+    ) -> AnyResult<Box<dyn InputEndpoint>>;
+}
+
+impl dyn InputTransport {
+    /// Lookup input transport by name.
+    pub fn get_transport(name: &str) -> Option<&'static dyn InputTransport> {
+        INPUT_TRANSPORT.get(name).map(|f| &**f)
+    }
+}
+
+/// Input transport endpoint receives a stream of bytes via the underlying
+/// data transport protocol and pushes it to the associated [`InputConsumer`].
+pub trait InputEndpoint: Send {
+    /// Pause the endpoint.
+    ///
+    /// The endpoint must stop pushing data downstream.  This method may
+    /// return before the dataflow has been fully paused, i.e., few additional
+    /// data buffers may be pushed downstream before the endpoint goes
+    /// quiescnent.
+    fn pause(&self) -> AnyResult<()>;
+
+    /// Start or restart the endpoint.
+    ///
+    /// The endpoint must start receiving data and pushing it downstream.
+    fn start(&self) -> AnyResult<()>;
+
+    /// Disconnect the endpoint.
+    ///
+    /// Disconnect the endpoint and stop receiving data.  This is the last
+    /// method invoked before the endpoint object is dropped.  It may return
+    /// before the dataflow has been fully terminated, i.e., few additional
+    /// data buffers may be pushed downstream before the endpoint goes
+    /// quiescnent.
```

Comment:
`quiescent` again here but should probably just be "quiet" or "before the endpoint disconnects"